### DISCUSSION TITLE
"decode_secrets: Error opening environment file '/proc/sys/urlader/environment'."

### DISCUSSION THREAD

**Discussion by BlackMage2** [PeterPawn/decoder#28] [General]
the decoder cant read a export file from a FB7530 164.07.21. Why? And where i can specify the password for the export file?
when i run "./decoder.x86_64 decode_secrets  ../../../../Downloads/FRITZ.Box\ 7530\ 164.07.21_13.05.22_1650.export" , reports the script:
`decode_secrets: Error opening environment file '/proc/sys/urlader/environment'.`

**Comment by PeterPawn** [comment]
Try `decoder --help` first.

There you should find the line

```
decode_export         - decrypt a FRITZ!OS export file
```

and taking this hint into account it's obviously, that you've used a wrong call.

Now try `decoder decode_export --help` ... this will reveal the "secret":

```
Usage:

    decode_export [ options ] [ -- ] [ password | serial maca ] [ [ < ] input-file ]

Supported options are:

-t, --tty              - don't quit execution, if STDIN is connected to a terminal device
-a, --alt-env filename - use an alternative source for the 'urlader environment'
-c, --checksum         - re-compute (and replace) the checksum for the provided export file, after
                         the cipher-text values were replaced with the corresponding clear-text
-l, --low-memory       - do not try to consolidate input data into a single buffer
-b, --block-size size  - read input data in blocks of the specified size
-v, --verbose          - show additional information on STDERR
-q, --quiet            - suppress any error messages on STDERR
-s, --strict           - treat warnings as errors and exit
-h, --help             - show this information (on STDOUT) and exit
-V, --version          - show version (on STDOUT) and exit
```

(and more explanations may be found there)

HTH

PS: Due to this is not really an "issue" (this means "a problem with the software", not "I don't know, how to use it."), ~so~ I'll convert this to a discussion thread.

**Reply by BlackMage2** [reply]
ohh, okay thanks

**Comment by terriblefire** [comment]
It would be really helpful to have some of this on the readme or in docs.

**Reply by PeterPawn** [reply]
Each applet has its own help screen, describing how to call it.

Any applet, which supports the `--alt-env` option, contains a description of this alternative, too, in its help screen.

Example:
```
vidar:~ $ decoder decode_secrets --help

decode_secrets, version 0.4

This program is a part of the project from https://github.com/PeterPawn/decoder.

Copyright (C) 2014-2020 P.Haemmerlein (peterpawn@yourfritz.de)

This project is free software, you can redistribute it and/or modify it under the terms of the GNU
General Public License as published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.

This project is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License under http://www.gnu.org/licenses/gpl-2.0.html for more details.

Purpose:

This program decrypts all occurrences of encrypted data on STDIN (if decryption is possible) and
writes the data to STDOUT, while replacing cipher-text with the corresponding clear-text values.

Usage:

    decode_secrets [ options ] [ -- ] [ parameter [ ... ] ] [ [ < ] input-file ]

Supported options are:

-t, --tty              - don't quit execution, if STDIN is connected to a terminal device
-a, --alt-env filename - use an alternative source for the 'urlader environment'
-l, --low-memory       - do not try to consolidate input data into a single buffer
-b, --block-size size  - read input data in blocks of the specified size
-v, --verbose          - show additional information on STDERR
-q, --quiet            - suppress any error messages on STDERR
-s, --strict           - treat warnings as errors and exit
-h, --help             - show this information (on STDOUT) and exit
-V, --version          - show version (on STDOUT) and exit

The input-file may be any text file, starting from a real TFFS node up to a single line of Base32
encoded cipher-text, as long as any encrypted value starts with the four dollar-signs. This can be
useful, if you extract some settings from a configuration file with other utilities (e.g.
'ar7cfgctl') and want to decode only a single value or a smaller amount of values.

STDIN has to be redirected or the program will be aborted - it's a protection against accidental
calls without proper redirection, otherwise the program would block the caller. If you really want
to call it with STDIN connected to a terminal device, specify the '--tty' option for the call.

If STDIN is connected to a terminal device and the '--tty' option (or '-t') wasn't specified, the
last expected value from command line is checked, wether it's the name of a readable file. In this
case the specified input-file will be used as source of input data.

The number of parameters depends on the mode of the call. If no parameter was specified, the
decryption key is built from device properties in the 'urlader environment'. If the program isn't
called on a FRITZ!OS device, the 'urlader environment' may be substituted by a text file. Its
location can be specified with the '--alt-env' (or '-a') option.

If a single parameter is used, it has to be a hexadecimal string representing the key to be used
for decryption.

If more than one parameter exists, it's an alternative approach to decrypt data from another
FRITZ!OS device with knowledge of its properties - this one doesn't need an environment file
from somewhere. For this mode, you need at least the 'SerialNumber', 'maca' and 'wlan_key' values
of the foreign device. If it has a pre-configured CWMP account, you need the value from the field
'tr069_passphrase' of the device too.

If you've got a copy of the 'urlader environment' for the device, where your input-file was
created, you should use this file with the '--alt-env' option instead of specifying three and more
values for the call.

The option '--low-memory' (or '-l') may be used, if your system has not enough free memory to hold
the input data (from really huge files) twice in memory for a short time.

The option '--block-size' (or '-b') to specify the input block size should seldom be necessary. If
you specify it, you can set a buffer size between 32 bytes and 16 Mbytes. Any valid number may be
followed by a 'K' or 'M' (upper or lower case) to use a factor of 2**10 or 2**20 with the given
number. This disables at the same time all further tries to consolidate the input data in a single
buffer (as if the option '--low-memory' was used).

vidar:~ $
```

In my opinion this is the best solution - combined with the explanations from `README.md`, which properties are required to be known for a successful decryption, EACH user should be able to find info about the required data and how to use it to call the needed function.

---

Or what else did you mean? Your comment wasn't really helpful with this - the original poster used the wrong function to decode an `export` file (`decode_secrets` instead of `decode_export`) and the hints from my answer above (call it with `decoder --help` first, read the output and use the RIGHT applet for the job) are REALLY obvious and common for any "command line utility". 

Further info WAS written down in the `README.md` file and much more explanations may be found in the mentioned bulletin board - that its only available in German shouldn't be any problem. In the past I'd beg some other people (who didn't write the article) to read it with an online translation tool (e.g. DeepL) in English and they found it sufficiently clear and comprehensible.

So if you want more explanations to be added anywhere, please be more specific, what's missing in your opinion.

**Comment by terriblefire** [comment]
Its your app so your rules but i disagree. there are too many applets and when you are unfamiliar with the application its a wall of text to try and parse to find what you want. A short summary of the applets and what they do in the README would be massively useful. If for nothing else than to be able to determine whether the application is going to be useful to you without having to download and compile it.

**Comment by PeterPawn** [comment]
> to determine whether the application is going to be useful to you without having to download and compile it

What's wrong with the documented purposes here (https://github.com/PeterPawn/decoder?tab=readme-ov-file#purposes):

```
Purpose(s):

    - decode encrypted credentials from any configuration file for system extensions contained in projects like Freetz or my own modfs
    - decode data from an exported settings file, where the export password is known
    - decode internal files from a foreign device, e.g. extracted from a TFFS dump
    - split export files into the contained settings files
    - recompute and check/change the CRC32 checksum at the very end of export files
    - encode/decode Base32 (with AVM's character set), Base64 and hexadecimal representations from/to raw binary data (the BusyBox project provides only a Base64 implementation)

Encoding of values is not provided yet - AVM's components accept clear-text values in nearly all places, where an encrypted value may be used.
```

And is this really the claimed

>  wall of text to try and parse to find what you want

? 

It's the second paragraph in this file, while the title at the beginning states explicitly:

> Decode secrets from/for AVM's FRITZ!OS-based devices

I can't see any reason (at least not a good one, until you name it), why the used applet names (the names are configurable, too) and/or their descriptions should be clearer than the already mentioned/documented purposes. Those purposes cover practically every (complex) function of this little utility - the other applets only exist, to provide access to internally used helper functions as an additional benefit.

**Comment by terriblefire** [comment]
I am just giving the feedback i found from my experience of using the tool. Since i probably wont need to use it again it really doesnt bother me. I just found finding what i wanted difficult for a simple one time decode of an export use case.

**Reply by PeterPawn** [reply]
> I just found finding what i wanted difficult 

I don't understand, what you mean. You wanted to decide, whether this project would be useful (for yourself) to decode an existing export file. If you've read the `README.md` file (thoroughly!), you should have gotten the first impression, whether it would/should cover your needs or not (and already in the 2nd paragraph).

If you aren't using any unusual platform, you may take one of the pre-compiled binaries (https://github.com/PeterPawn/decoder/tree/master/bin) - otherwise it takes only 5 minutes to compile an own one (if the prerequisites are met).

Your feedback is really welcome, but your arguments can't convince me. Why should knowledge of the contained applet names and their descriptions (and as mentioned already, these names are configurable at compile-time, just like the applets to include) be better than the existing description of the purposes of this project?

If you want/need only a "one-time solution" and are really deterred by my "wall of text", you'd better use another implementation: https://www.mengelke.de/Projekte/FritzBox-JSTool.html (the link is for the offline version) - but there you have to read the documentation first, too.
